### ISSUE TITLE
look to this: how KDE want to circumvent Wayland

### ISSUE THREAD

**Issue by rilian-la-te** [lestcape/unity-gtk-module#2]
I rewrite it in Vala, but can provide original XML:

```Vala
    [DBus (name = "org.kde.kappmenu")]
    public interface KDEAppMenu : DBusProxy
    {
        public abstract void show_menu(int x,int y,string service, ObjectPath path,int actionId);
        public abstract void reconfigure();
        public signal void reconfigured();
        public signal void show_request(string service, ObjectPath path,int actionId);
        public signal void menu_shown(string service, ObjectPath path);
        public signal void menu_hidden(string service, ObjectPath path);
    }
```

They want to force each window to do this. We also can (with exception to be a MenuModel instead, not DBusMenu). And no compositor envolved AFAIK)))

**Comment by lestcape** [comment]
But for this skeleton, i think this is for the appindicator menu and not for the menubar. Not?

I don't see a reference to the owner window, but in the appindicator you don't need a reference. Just the x,y coordinate where the popupmenu will be displayed.

**Comment by rilian-la-te** [comment]
For menubar, kde guys does not have an appmenu. We do not need a reference for window in this case. Just collect all windows that have this interface exported in standard object path (KDE have /KAppMenu) and connect to siginals.

**Comment by lestcape** [comment]
Not, appindicator and appmenu are not the same things. The appindicator is the replace from the old systray. Kde have the appindicator support and this is the Dbus structure of the appindicators.

Gnome have not appindicator support by default and also not systray support right now, but ubuntu have an extension to support appindicators and also there are other extensions to support systray.

**Comment by lestcape** [comment]
In any case, we need to know the active window and what menu display in what case. The appindicators not need that because all appindicators are displayed at the same time, not matter the active window. This is why you don't need to know the owner window in the appindicator case, it's not useful.

**Comment by lestcape** [comment]
If you want to know how menubar is supported inside kde, i will tell you, is using the compisitor. The kwin compositor. Relevant commit about wayland support is here: https://cgit.kde.org/kwin.git/commit/?id=93938d60b84df81733da7b3f697332fc76fe085b

**Comment by lestcape** [comment]
They use as id the `sender + path` to find a menu and distinguish between what they called AbstractClient.  I think abstract clients are an abstract window representation and the menubar is only possible for windows that implement or can be implement the AbstractClient class.

See: **findAbstractClientWithApplicationMenu**

**Comment by lestcape** [comment]
To be honest what i don't know is how we can support qt wayland windows yet (in gnome they don't exist). I don't think gnome will wrap all properties of a qt window and we need this properties to extract the sender and the path from the WM/Composuitor otherwise we will have not a way to get it.

**Comment by rilian-la-te** [comment]
They have also a non-compositor approach in Plasma. I will tell you tomorrow. And you are right about abstract client. I will look for it.

**Comment by lestcape** [comment]
Ok, thanks. I really just have ideas about what i see... And from what i see, is a big problem to us really. Not now, but in the future...

Is nice if there are a non compositor approach, because otherwise, there are not solution to us then.

**Comment by lestcape** [comment]
Do you not need to start the appmenu service after the mate-session.service? https://github.com/rilian-la-te/vala-panel-appmenu/blob/master/subprojects/appmenu-gtk-module/data/appmenu-gtk-module.service.in#L4

**Comment by rilian-la-te** [comment]
Yes, thanks.

**Comment by lestcape** [comment]
Do you know, if i try to rename the package in ubuntu to appmenu-gtk-module it will enter in conflict with unity-gtk-module. To install it I will need to uninstall unity-gtk-module. I suspect it will have consequence then with Unity desktop. So, i don't know if rename the package for unity will be the best solution...

**Comment by rilian-la-te** [comment]
Why it will be conflicted?

**Comment by lestcape** [comment]
I think because of this: https://github.com/lestcape/unity-gtk-module/blob/ad9bbdf95b8ff297f38524efb511fb9b5a6d9da1/debian/control#L117

**Comment by lestcape** [comment]
Both package replace he same package, so it will enter in conflict.

**Comment by rilian-la-te** [comment]
It is strange by me, because it can be a valid case.

**Comment by lestcape** [comment]
Maybe remove the replace, or set the replace and the conflict with unity-gtk-module itself?

**Comment by rilian-la-te** [comment]
I cannot find exact commit for now, but they tries to connect by Registrar to this interface on each registered window (do not care about id), and then monitor a signals like showMenu and hideMenu.
I think, we can also try this approach on unity-gtk-module, but we need to send MenuModels.
I will try to write platform function for this later.

**Comment by rilian-la-te** [comment]
I think, just replace is enough. Because no conflict involved by fact (AFAIK I have no conflicted files with original unity-gtk-module)

**Comment by lestcape** [comment]
@rilian-la-te I underestand the aproach, but not how to know what Dbus service Show/Hidde we will need to trigger.

**Comment by lestcape** [comment]
> AFAIK I have no conflicted files with original unity-gtk-module

The conflict is declared as a Debian package conflict by functionality, not as a real source code conflict. This is why i say remove this declaration will help...

**Comment by rilian-la-te** [comment]
Window AFAIK can know its state on wayland (and do not know about state if other windows). So, it can inform Registrar about this state by emitting showMenu / hideMenu on this interface.

**Comment by rilian-la-te** [comment]
Feel free to remove (but consult with appmenu-gtk-module maintainer in Debian, I am not a Debian-based user)

**Comment by lestcape** [comment]
> Window AFAIK can know its state on wayland (and do not know about state if other windows). So, it can inform Registrar about this state by emitting showMenu / hideMenu on this interface.

Yes and again the point of how associate the dbus path and sender with a real window (the active window).

Feel free to remove (but consult with appmenu-gtk-module maintainer in Debian, I am not a Debian-based user)

It's not in your code, it's in the debian folder (of the debian package) that evolved your source to be possible install it with dh and then upload it to launchpad.

**Comment by lestcape** [comment]
Remember y kde you can asociate the Dbus path and sender of a window with the window, because kwin allow you to do that, but mutter don't allow you to do that or any other WM implementation, at less not for now. In wayland clients are isolate (for security reason), so is in my opinion difficult that this can be shared easy.

**Comment by rilian-la-te** [comment]
It should live in vala-project-packages.

We do not need to know active window. We think than active menu is last emiiter of showMenu signal if it does not emitted hideMenu yet.

**Comment by rilian-la-te** [comment]
And if in module we will emit showMenu on each realize, and hideMenu on each unrealize, I think this approach will work without compositor.

**Comment by lestcape** [comment]
A window realize or unrealized is not directly associate with the active window at all, but i get your idea and i think it will work.

**Comment by lestcape** [comment]
I can work on ubuntu with your source easy now:
https://github.com/lestcape/appmenu-gtk-module

So, i will work to make possible use your appmenu-gtk-module inside the gnome-global-appmenu extension.

**Comment by rilian-la-te** [comment]
And please, commit all your work to upstream (debian folder too, it should go to vala-project-packages)

**Comment by rilian-la-te** [comment]
And why autotools? I switched it to CMake a whole ago)

**Comment by lestcape** [comment]
Because like i say i don't know anything of how to package a thing, so i used what i have and what i know is working on launchpad.

CMake will work on launchpand? I don't know and i don't know how if it's possible.

**Comment by rilian-la-te** [comment]
Yes, it works) I has a repo in that, when you can see a very out date version of my panel and plugin)

**Comment by lestcape** [comment]
Please give me a reference to this package then. I can try...

**Comment by rilian-la-te** [comment]
You can find it in https://launchpad.net/~rilian-la-te/+archive/ubuntu/vala-panel-plugins , but it is very outdated.
You can take official vala-panel or vala-panel-appmenu package in 17.10.

**Comment by lestcape** [comment]
I see what i need to do to port it from autotools to CMake and it's a lot to me. Also i don't know why you want that i port it to CMake.

For another hand, you say:

> And please, commit all your work to upstream (debian folder too, it should go to vala-project-packages)


Are you sure of that? Martin request to you the opposite: 

> Can we drop the debian/ from your git repository? 

https://github.com/rilian-la-te/vala-panel-appmenu/issues/53#issuecomment-291024244

**Comment by rilian-la-te** [comment]
Martin ask me to remove debian folder from vala-panel-appmenu project. But vala-project-packages is especially for packages.
You find packaging CMake project difflicult?
Or porting a module?
If porting a module - module is already ported. It is completely buildable without vala-panel-appmenu.

**Comment by lestcape** [comment]
Is that: i never play with CMake in my life... For another hand, you have not support at less for docs and autopilot using CMake. You have support for rpm and other vala stuff that are not needed (maybe just some of then for the examples).

What is the idea of port a complete thing to CMake to then be incomplete? So, i will need to also add support for the missing piece and i don't know how.

**Comment by rilian-la-te** [comment]
I will add support of docs and autopilot, but for me it not needed.

For me there is another thing - I do not know autotools at all, but know CMake a little. So, for me CMake is better.

**Comment by rilian-la-te** [comment]
RPM is needed)

**Comment by lestcape** [comment]
> RPM is needed)

Not for me, i just want to support ubuntu for my personal interest in other distros they have your package.

**Comment by rilian-la-te** [comment]
Can you use meson then?

**Comment by lestcape** [comment]
meson is new, and it's the gnome new way to package the things. I will need to learn it.

To be honest, i don't like to be a maintainer of a package or something else. Just i want to use your code in ubuntu, to not need to work in a virtual machine and don't break the extension in the Ubuntu LTS release.

I will not support all ubuntu releases, i don't want to be a maintainer of a package, so thats only to show it's possible...
